File: src/utils/chefIntro.js
import {
  DESSERT_STYLE_VARIANTS,
  buildDessertDishTitle,
  pickPrimaryFlavorLabel,
} from './dessertDishTitle'
import { buildGroundedChefTitle } from './recipeGrounding'
import { canonicalIngredient } from '../data/ingredientKnowledge'
import { getIngredientLabel } from '../data/ingredientLabels'

const QTY_PREFIX =
  /^[\d./]+\s*(?:כפ(?:ית|ות)|כ(?:ף|פות)|גרם|מ"ל|כוס(?:ות)?|tsp|tbsp|gram|grams|g|ml|cup|cups)?\s*/i

function stripQty(raw) {
  return String(raw ?? '')
    .replace(QTY_PREFIX, '')
    .trim()
}

function mainCanon(ingredients = []) {
  return ingredients
    .map((item) => canonicalIngredient(stripQty(item)))
    .filter(Boolean)
}

function uniqueNames(items = []) {
  return [...new Set(items.filter(Boolean))]
}

export function suggestDishOptions(ingredients = [], { language = 'he', recipeType = 'meal' } = {}) {
  if (recipeType === 'dessert') {
    const main = pickPrimaryFlavorLabel(ingredients, language)
    return uniqueNames(
      DESSERT_STYLE_VARIANTS.map((variant) =>
        language === 'he' ? variant.titleHe(main) : variant.titleEn(main),
      ),
    ).slice(0, 4)
  }

  const canon = new Set(mainCanon(ingredients))
  const flavor = pickPrimaryFlavorLabel(ingredients, language)

  if (canon.has('cinnamon') && (canon.has('cream') || canon.has('milk'))) {
    return language === 'he'
      ? ['קרם קינמון חם', 'פודינג קינמון', 'רוטב קינמון מתוק']
      : ['Warm cinnamon cream', 'Cinnamon pudding', 'Sweet cinnamon sauce']
  }
  if (canon.has('chicken')) {
    return language === 'he'
      ? ['עוף במחבת', 'עוף בתנור', 'מרק עוף קל']
      : ['Pan-seared chicken', 'Oven-baked chicken', 'Light chicken soup']
  }
  if (canon.has('pasta')) {
    return uniqueNames([
      buildGroundedChefTitle(ingredients, ingredients, language),
      language === 'he' ? 'פסטה מהירה' : 'Quick pasta',
    ]).slice(0, 3)
  }
  if (canon.has('tomato') && (canon.has('egg') || canon.has('eggs'))) {
    return language === 'he'
      ? ['שקשוקה', 'חביתת עגבניות', 'ביצים ברוטב עגבניות']
      : ['Shakshuka', 'Tomato omelette', 'Eggs in tomato sauce']
  }
  if (canon.has('rice')) {
    return uniqueNames([
      buildGroundedChefTitle(ingredients, ingredients, language),
      language === 'he' ? 'אורז ביתי' : 'Homestyle rice',
    ]).slice(0, 3)
  }

  const label = flavor || getIngredientLabel([...canon][0] ?? 'ingredient', language)
  return language === 'he'
    ? [`${label} במחבת`, `תבשיל ${label}`, `מנה חמה עם ${label}`]
    : [`${label} skillet`, `${label} stew`, `Warm ${label} dish`]
}

function buildChoiceReason(chosenName, options, { language = 'he', cookingTime = 30 } = {}) {
  const chosen = String(chosenName ?? '').trim()
  if (!chosen) {
    return language === 'he'
      ? 'בחרתי מנה שמתאימה למרכיבים שיש לך.'
      : 'I picked a dish that fits what you have on hand.'
  }

  if (language === 'he') {
    if (options.length <= 1) {
      return `${chosen} מתאימה במיוחד למרכיבים שיש לך — פשוטה, טעימה, ובזמן של כ-${cookingTime} דקות.`
    }
    return `האפשרות הכי פשוטה וטעימה היא ${chosen} — מתאימה לזמן של כ-${cookingTime} דקות ולמרכיבים שיש לך.`
  }

  if (options.length <= 1) {
    return `${chosen} is a great fit for your ingredients — simple, tasty, and ready in about ${cookingTime} minutes.`
  }
  return `The simplest and most appealing option is ${chosen} — it fits your ingredients and about ${cookingTime} minutes of cooking time.`
}

/**
 * Friendly chef intro: alternatives + why this recipe was chosen.
 */
export function buildChefIntro(
  ingredients = [],
  {
    chosenName = '',
    language = 'he',
    recipeType = 'meal',
    cookingTime = 30,
  } = {},
) {
  let options = suggestDishOptions(ingredients, { language, recipeType })
  if (chosenName && !options.includes(chosenName)) {
    options = [chosenName, ...options.filter((item) => item !== chosenName)]
  } else if (chosenName) {
    options = [chosenName, ...options.filter((item) => item !== chosenName)]
  }
  options = uniqueNames(options).slice(0, 4)

  if (options.length === 0 && chosenName) {
    options = [chosenName]
  }

  const intro =
    language === 'he'
      ? 'עם המרכיבים שיש לך אפשר להכין:'
      : 'With the ingredients you have, you could make:'
  const bullets = options.map((item) => `• ${item}`).join('\n')
  const reason = buildChoiceReason(chosenName || options[0], options, { language, cookingTime })

  return `${intro}\n${bullets}\n\n${reason}`
}

export function buildChefIntroForRecipe(recipe, { language = 'he', recipeType = 'meal', cookingTime = 30 } = {}) {
  return buildChefIntro(recipe?.ingredients ?? [], {
    chosenName: recipe?.name ?? '',
    language,
    recipeType,
    cookingTime,
  })
}

export function enrichDescriptionWithChefIntro(
  description,
  ingredients = [],
  {
    chosenName = '',
    language = 'he',
    recipeType = 'meal',
    cookingTime = 30,
    hasUserIngredients = true,
  } = {},
) {
  if (!hasUserIngredients || !ingredients.length) {
    return description
  }
  const intro = buildChefIntro(ingredients, { chosenName, language, recipeType, cookingTime })
  return intro
}

File: backend/chef_intro.py
"""Friendly chef-style intro: dish options + why this recipe was chosen."""

from __future__ import annotations

import re

from dessert_dish_title import DESSERT_STYLE_VARIANTS, pick_primary_flavor_label
from ingredient_relevance import canonical_ingredient

QTY_PREFIX = re.compile(
    r"^[\d./]+\s*(?:כפ(?:ית|ות)|כ(?:ף|פות)|גרם|מ\"ל|כוס(?:ות)?|tsp|tbsp|gram|grams|g|ml|cup|cups)?\s*",
    re.IGNORECASE,
)


def _strip_qty(raw: str) -> str:
    return QTY_PREFIX.sub("", (raw or "").strip()).strip()


def _main_canon(ingredients: list[str]) -> list[str]:
    return [canonical_ingredient(_strip_qty(item)) or item for item in ingredients]


def _unique(items: list[str]) -> list[str]:
    seen: set[str] = set()
    result: list[str] = []
    for item in items:
        if item and item not in seen:
            seen.add(item)
            result.append(item)
    return result


def suggest_dish_options(
    ingredients: list[str],
    *,
    language: str = "he",
    recipe_type: str = "meal",
    category: str = "dairy",
) -> list[str]:
    from recipe_dish_patterns import get_dish_suggestions

    suggestions = get_dish_suggestions(
        ingredients,
        language=language,
        recipe_type=recipe_type,
        category=category,
    )
    if suggestions:
        return suggestions

    if recipe_type == "dessert":
        main = pick_primary_flavor_label(ingredients, language)
        options = [
            variant["title_he"](main) if language == "he" else variant["title_en"](main)
            for variant in DESSERT_STYLE_VARIANTS
        ]
        return _unique(options)[:4]

    canon = set(_main_canon(ingredients))
    flavor = pick_primary_flavor_label(ingredients, language)

    if "cinnamon" in canon and ("cream" in canon or "milk" in canon):
        return (
            ["קרם קינמון חם", "פודינג קינמון", "רוטב קינמון מתוק"]
            if language == "he"
            else ["Warm cinnamon cream", "Cinnamon pudding", "Sweet cinnamon sauce"]
        )
    if "chicken" in canon:
        return (
            ["עוף במחבת", "עוף בתנור", "מרק עוף קל"]
            if language == "he"
            else ["Pan-seared chicken", "Oven-baked chicken", "Light chicken soup"]
        )
    if "pasta" in canon:
        return (
            ["פסטה ברוטב שמנת", "פסטה עם ירקות", "פסטה מהירה"]
            if language == "he"
            else ["Creamy pasta", "Vegetable pasta", "Quick pasta skillet"]
        )
    if "tomato" in canon and ("egg" in canon or "eggs" in canon):
        return (
            ["שקשוקה", "חביתת עגבניות", "ביצים ברוטב עגבניות"]
            if language == "he"
            else ["Shakshuka", "Tomato omelette", "Eggs in tomato sauce"]
        )
    if "rice" in canon:
        return (
            ["אורז מוקפץ", "תבשיל אורז", "אורז עם ירקות"]
            if language == "he"
            else ["Fried rice", "Rice pilaf", "Vegetable rice"]
        )

    label = flavor or "מנה"
    return (
        [f"{label} במחבת", f"תבשיל {label}", f"מנה חמה עם {label}"]
        if language == "he"
        else [f"{label} skillet", f"{label} stew", f"Warm {label} dish"]
    )


def _build_choice_reason(
    chosen_name: str,
    options: list[str],
    *,
    language: str = "he",
    cooking_time: int = 30,
) -> str:
    chosen = (chosen_name or "").strip()
    if not chosen:
        return (
            "בחרתי מנה שמתאימה למרכיבים שיש לך."
            if language == "he"
            else "I picked a dish that fits what you have on hand."
        )

    if language == "he":
        if len(options) <= 1:
            return (
                f"{chosen} מתאימה במיוחד למרכיבים שיש לך — "
                f"פשוטה, טעימה, ובזמן של כ-{cooking_time} דקות."
            )
        return (
            f"האפשרות הכי פשוטה וטעימה היא {chosen} — "
            f"מתאימה לזמן של כ-{cooking_time} דקות ולמרכיבים שיש לך."
        )

    if len(options) <= 1:
        return (
            f"{chosen} is a great fit for your ingredients — "
            f"simple, tasty, and ready in about {cooking_time} minutes."
        )
    return (
        f"The simplest and most appealing option is {chosen} — "
        f"it fits your ingredients and about {cooking_time} minutes of cooking time."
    )


def build_chef_intro(
    ingredients: list[str],
    *,
    chosen_name: str = "",
    language: str = "he",
    recipe_type: str = "meal",
    cooking_time: int = 30,
) -> str:
    options = suggest_dish_options(ingredients, language=language, recipe_type=recipe_type)
    if chosen_name:
        options = _unique([chosen_name, *[item for item in options if item != chosen_name]])
    options = options[:4]
    if not options and chosen_name:
        options = [chosen_name]

    intro = (
        "עם המרכיבים שיש לך אפשר להכין:"
        if language == "he"
        else "With the ingredients you have, you could make:"
    )
    bullets = "\n".join(f"• {item}" for item in options)
    reason = _build_choice_reason(chosen_name or (options[0] if options else ""), options, language=language, cooking_time=cooking_time)
    return f"{intro}\n{bullets}\n\n{reason}"
